inputNormalizeEdgesTransparent blur isn't available on all platforms.

mattwoodrow opened the pull request:
#### 85a241ca1e5f23b30574ccfa51d54f17b6e3932a
<pre>
inputNormalizeEdgesTransparent blur isn&apos;t available on all platforms.
<a href="https://bugs.webkit.org/show_bug.cgi?id=273280">https://bugs.webkit.org/show_bug.cgi?id=273280</a>
&lt;<a href="https://rdar.apple.com/126606399">rdar://126606399</a>&gt;

Reviewed by Darin Adler.

We currently try to set this everywhere, but the key isn&apos;t always available.

We should use inputHardEdges instead, but since this looks worse, we should try to use
inputNormalizeEdges when we know the blurred area is opaque.

This detects when the layer that establishes a backdrop root also has an opaque
background color, and passes that down to descendants in order to pick the right blur
mode.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::updateBackdropFilters):
(WebCore::GraphicsLayerCA::updateBackdropRoot):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCAFilters.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
(WebCore::PlatformCAFilters::setFiltersOnLayer):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::setBackdropRootIsOpaque):
(WebCore::PlatformCALayerCocoa::setFilters):
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
(WebKit::RemoteAcceleratedEffectStack::applyEffectsFromMainThread const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::backdropRootIsOpaque const):
(WebKit::RemoteLayerTreeNode::setBackdropRootIsOpaque):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setBackdropRootIsOpaque):

Canonical link: <a href="https://commits.webkit.org/278155@main">https://commits.webkit.org/278155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4948fb7642931876cfca9547b25dc760963eec81

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49646 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28934 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52888 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/322 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51951 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26551 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40510 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51746 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26466 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21628 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23911 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43945 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8017 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44449 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54465 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24733 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20896 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47885 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26000 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10897 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26848 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25725 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
```diff
@@ -646,6 +647,17 @@ - (void)setOwner:(WebCore::PlatformCALayer*)owner
+}
+
+
+void PlatformCALayerCocoa::setBackdropRootIsOpaque(bool backdropRootIsOpaque)
```
Extra blank line here.

darinadler commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
```diff
@@ -327,16 +327,15 @@ static unsigned keyValueCountForFilter(const FilterOperation& filterOperation)
+                if (backdropIsOpaque)
+                    [filter setValue:@YES forKey:@"inputNormalizeEdges"];
+                else {
 #if PLATFORM(VISION)
```
Would be nice if this was a `HAVE` or `USE` instead of `PLATFORM`; I'm guessing some day we might have this filter on other platforms.

darinadler replied:
I don’t mean you need to do this now.

darinadler commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
```diff
@@ -120,6 +120,8 @@ class PlatformCALayerRemote : public WebCore::PlatformCALayer, public CanMakeWea
 
     void setIsBackdropRoot(bool) override;
+    bool backdropRootIsOpaque() const override;
+    void setBackdropRootIsOpaque(bool) override;
```
Can these be `final` instead?

mattwoodrow replied:
They can, but the file-local style is to use `override`.

Feels a bit weird to me to have just these new functions differ, but I'm not strongly attached to that opinion.

darinadler replied:
Project style is to use `final` whenever it can be used instead of `override`. Ideally we don’t start developing different file-local styles; I try to resist that when I can.

darinadler commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
```diff
@@ -88,6 +88,8 @@ class PlatformCALayerCocoa final : public PlatformCALayer {
 
     void setIsBackdropRoot(bool) override;
+    bool backdropRootIsOpaque() const override;
+    void setBackdropRootIsOpaque(bool) override;
```
Can these be final instead?

darinadler commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -2521,8 +2524,13 @@ void GraphicsLayerCA::updateBackdropFilters(CommitState& commitState)
-    if (!(m_uncommittedChanges & BackdropFiltersChanged))
+    if (!(m_uncommittedChanges & BackdropFiltersChanged)) {
+        // Opaque state depends on ancestor state, and is cheap to set, so
+        // just conditionally update it.
```
Would prefer a long line for this comment instead of two lines.

darinadler commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -2533,12 +2541,14 @@ void GraphicsLayerCA::updateBackdropFilters(CommitState& commitState)
 
     if (m_layerClones) {
         for (auto& clone : m_layerClones->backdropLayerClones) {
             PlatformCALayer* cloneLayer = clone.value.get();
```
Could have been done like this:
```
for (auto& cloneLayer : m_layerClones->backdropLayerClones.values())
```

darinadler commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -270,8 +270,11 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
+
     if (properties.changedProperties & LayerChange::FiltersChanged)
-        PlatformCAFilters::setFiltersOnLayer(layer, properties.filters ? *properties.filters : FilterOperations(), layerTreeHost->cssUnprefixedBackdropFilterEnabled());
+        PlatformCAFilters::setFiltersOnLayer(layer, properties.filters ? *properties.filters : FilterOperations(), layerTreeNode ? layerTreeNode->backdropRootIsOpaque() : false);
```
I prefer: `layerTreeNode && layerTreeNode->backdropRootIsOpaque()`

mattwoodrow committed:
inputNormalizeEdgesTransparent blur isn't available on all platforms.
https://bugs.webkit.org/show_bug.cgi?id=273280
<rdar://126606399>

Reviewed by Darin Adler.

We currently try to set this everywhere, but the key isn't always available.

We should use inputHardEdges instead, but since this looks worse, we should try to use
inputNormalizeEdges when we know the blurred area is opaque.

This detects when the layer that establishes a backdrop root also has an opaque
background color, and passes that down to descendants in order to pick the right blur
mode.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::updateBackdropFilters):
(WebCore::GraphicsLayerCA::updateBackdropRoot):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCAFilters.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
(WebCore::PlatformCAFilters::setFiltersOnLayer):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::setBackdropRootIsOpaque):
(WebCore::PlatformCALayerCocoa::setFilters):
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
(WebKit::RemoteAcceleratedEffectStack::applyEffectsFromMainThread const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::backdropRootIsOpaque const):
(WebKit::RemoteLayerTreeNode::setBackdropRootIsOpaque):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setBackdropRootIsOpaque):

Canonical link: https://commits.webkit.org/278155@main
